agent/dns.go

```diff
@@ -1483,10 +1369,284 @@ func findWeight(node structs.CheckServiceNode) int {
 	}
 }
 
+// serviceNodeAddr is used to identify target service address
+func (d *DNSServer) serviceNodeAddr(serviceNode structs.CheckServiceNode, dc string, dnsQuery string) string {
+	addr := d.agent.TranslateAddress(dc, serviceNode.Node.Address, serviceNode.Node.TaggedAddresses)
+
+	if serviceNode.Service.Address != "" {
+		addr = serviceNode.Service.Address
+	}
+
+	// If the service address is a CNAME for the service we are looking
+	// for then use the node address.
+	if dnsQuery == strings.TrimSuffix(addr, ".")+"." {
+		addr = serviceNode.Node.Address
+	}
+
+	return addr
+}
+
+func (d *DNSServer) encodeIPAsFqdn(dc, ip string) string {
+	addr := net.ParseIP(ip)
+	ipv4 := addr.To4()
+	if ipv4 != nil {
+		ipStr := hex.EncodeToString(addr)
+		return fmt.Sprintf("%s.addr.%s.%s", ipStr[len(ipStr)-(net.IPv4len*2):], dc, d.domain)
+	} else {
+		return fmt.Sprintf("%s.addr.%s.%s", hex.EncodeToString(addr), dc, d.domain)
+	}
+}
+
+func makeARecord(qType uint16, addr string, ttl time.Duration) dns.RR {
+	ip := net.ParseIP(addr)
```

Comment:
Do we have some guarantee here that this is called only when we already know the addr is a valid IP address?

If so could you add a comment about that.

If that is not the case then we should add a 

```
if ip == nil {
   return nil
}

Reply:
Replace string input by net.IP